File: 1339.dart
import 'dart:math';

class TreeNode {
  int val;
  TreeNode? left;
  TreeNode? right;
  TreeNode([this.val = 0, this.left, this.right]);
}

class Solution {
  int nowRes = 0;

  int maxProduct(TreeNode? root) {
    int sums = treeSum(root!);
    dfs(root, sums);
    return nowRes.toInt() % (1000000000 + 7);
  }

  int dfs(TreeNode? root, int sumTree) {
    if (root == null) return 0;
    var lDfs = dfs(root.left, sumTree);
    var rDfs = dfs(root.right, sumTree);
    int nowSum = root.val + lDfs + rDfs;
    nowRes = max(nowRes, nowSum * (sumTree - nowSum));
    return nowSum;
  }

  //StackOverflow exception
  int treeSum(TreeNode? root) {
    if (root == null) return 0;
    return root.val + treeSum(root.left) + treeSum(root.right);
  }
}

File: 1339.py
class Solution:
    nowRes = 0

    def maxProduct(self, root: Optional[TreeNode]) -> int:
        treeSum(root)
        return self.dfs(root, treeSum(root)) % (1000000000 + 7)

    def dfs(self, root, sumTree):
        if root is None:
            return 0
        lDfs = self.dfs(root.left, sumTree)
        rDfs = self.dfs(root.right, sumTree)
        nowSum = root.val + lDfs + rDfs
        self.nowRes = max(self.nowRes, nowSum * (sumTree - nowSum))
        return nowSum


def treeSum(root) -> int:
    if root is None:
        return 0
    return root.val + treeSum(root.left) + treeSum(root.right)
